File: src/lowlevel/client/icefsoperators/icefsFsyncDir.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-03-30 04:23:15
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsFsyncDir.cpp
 * @Description: 
 * 
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsFsyncDir(fuse_req_t fuseReq, fuse_ino_t inode,
                                  int dataSync, struct fuse_file_info *fi) {
  IcefsFsyncDirReq req;
  IcefsFsyncDirRes res;
  grpc::ClientContext ctx;
  ICEFS_PR_FUNCTION;
  FuseReq *fuseReqToSend = new FuseReq();
  FuseCtx *fuseCtx = new FuseCtx();
  FuseFileInfo *fileInfo = new FuseFileInfo();
  IcefsFillFuseReq(fuseReqToSend, fuseCtx, fuseReq);
  req.set_allocated_req(fuseReqToSend);
  req.set_inode(inode);
  req.set_data_sync(dataSync);
  IcefsFillFuseFileInfoOut(fileInfo, fi);
  req.set_allocated_file_info(fileInfo);

  grpc::Status status = stub_->DoIcefsFsyncDir(&ctx, req, &res);
  if (status.ok() && !res.status()) {
    fuse_reply_err(fuseReq, ICEFS_EOK);
  } else {
    fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
    ICEFS_PR_ERR_STATUS;
  }
}

File: src/lowlevel/server/icefsoperators/icefsFsyncDir.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-17 16:44:07
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsFsyncDir.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

// TODO: Maybe not work
func (s *IcefsServer) doIcefsFsyncDir(dataSync int32, fh uint64) (status int32) {
	var fd int
	var err error

	s.dirCacheLock.RLock()
	dir := s.getIcefsDir(fh)
	if dir == nil {
		s.dirCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}
	dir.dirLock.RLock()
	s.dirCacheLock.RUnlock()
	fd = IcefsDirFd(dir.dirStream)
	dir.dirLock.RUnlock()

	if dataSync != 0 {
		err = syscall.Fdatasync(fd)
	} else {
		err = syscall.Fsync(fd)
	}
	status = icefserror.IcefsStdErrno(err)
	return
}

func (s *IcefsGRpcServer) DoIcefsFsyncDir(ctx context.Context, req *pb.IcefsFsyncDirReq) (*pb.IcefsFsyncDirRes, error) {
	var res pb.IcefsFsyncDirRes

	res.Status = s.server.doIcefsFsyncDir(req.DataSync, req.Fh)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsFsyncDir(ctx context.Context, req *icefsthrift.IcefsFsyncDirReq) (*icefsthrift.IcefsFsyncDirRes, error) {
	var res icefsthrift.IcefsFsyncDirRes

	res.Status = s.server.doIcefsFsyncDir(req.DataSync, uint64(req.Fh))

	return &res, nil
}
